tests/proxy/cp/count_down_latch_test.py

```diff
@@ -0,0 +1,168 @@
+        self.assertFalse(latch.await_latch(-1))
+
+    def test_await_latch_zero_timeout(self):
+        latch = self._get_latch(1)
+        self.assertFalse(latch.await_latch(0))
+
+    def test_await_latch_with_timeout(self):
+        latch = self._get_latch(1)
+        start = time.time()
+        self.assertFalse(latch.await_latch(0.1))
+        time_passed = time.time() - start
+        self.assertTrue(time_passed > 0.1)
+
+    def test_await_latch_multiple_waiters(self):
+        latch = self._get_latch(1)
+
+        completed = AtomicInteger()
+
+        def run():
+            latch.await_latch(inf)
+            completed.get_and_increment()
+
+        count = 10
+        threads = []
+        for _ in range(count):
+            t = Thread(target=run)
+            threads.append(t)
+            t.start()
+
+        latch.count_down()
+
+        def assertion():
+            self.assertEqual(count, completed.get())
+
+        self.assertTrueEventually(assertion)
+
+        for i in range(count):
+            threads[i].join()
+
+    def test_await_latch_response_on_count_down(self):
+        latch = self._get_latch()
+        self.assertTrue(latch.await_latch(inf))
+        self.assertTrue(latch.try_set_count(1))
+
+        # make a non-blocking request
+        future = latch._wrapped.await_latch(inf)
+        t = Thread(target=lambda: latch.count_down())
+        t.start()
+        t.join()
+
+        def assertion():
+            self.assertTrue(future.done())
+            self.assertTrue(future.result())
+
+        self.assertTrueEventually(assertion)
+
+    def test_count_down(self):
+        latch = self._get_latch(10)
+
+        for i in range(9, -1, -1):
+            self.assertIsNone(latch.count_down())
+            self.assertEqual(i, latch.get_count())
+
+    def test_count_down_retry_on_timeout(self):
+        latch = self._get_latch(1)
+
+        original = latch._wrapped._request_count_down
+        called_count = AtomicInteger()
+
+        def mock(expected_round, invocation_uuid):
+            if called_count.get_and_increment() < 2:
+                return ImmediateExceptionFuture(OperationTimeoutError("xx"))
+            return original(expected_round, invocation_uuid)
+
+        latch._wrapped._request_count_down = mock
+
+        latch.count_down()
+        self.assertEqual(3, called_count.get())  # Will resolve on it's third call. First 2 throws timeout error
+        self.assertEqual(0, latch.get_count())
+
+    def test_get_count(self):
+        latch = self._get_latch(1)
+        self.assertEqual(1, latch.get_count())
+        latch.count_down()
+        self.assertEqual(0, latch.get_count())
+        latch.try_set_count(10)
+        self.assertEqual(10, latch.get_count())
+
+    def test_try_set_count(self):
+        latch = self._get_latch()
+        self.assertTrue(latch.try_set_count(3))
+        self.assertEqual(3, latch.get_count())
+
+    def test_try_set_count_when_count_is_already_set(self):
```

Comment:
nit: does it make sense include a test that verifies that `try_set_count` succeeds once the count goes to zero?

Reply:
Good idea. I added the extra test case.